### PR TITLE
feat: select configurations by regexp and/or attributes

### PR DESCRIPTION
**Pull request by kyegupov**
- [x] Ready for review
- [x] Follows CONTRIBUTING rules
- [x] Reviewed by Snyk internal team

#### What does this PR do?

Note: this is an extension of https://github.com/snyk/snyk-gradle-plugin/pull/59

* replace the old pseudo-Gradle option `-- --configuration=foo` with `--configuration-matching=^foo$`
* (still support the legacy option if provided)
* instead of selecting one or all Gradle configurations, select many by regex
* if all the matching configurations have the same value of a certain attribute, use that in the merged configuration (so that Android matching rules can work)

#### Where should the reviewer start?

First see the PR https://github.com/snyk/snyk-gradle-plugin/pull/59, then see changes in index.ts and init.gradle

#### How should this be manually tested?

Test projects with various configurations and compare snyk output with that of `gradle dependencies`

#### Tickets

https://snyksec.atlassian.net/browse/BST-571

#### Screenshot (suggestion in the error message)
image

### CODE REVIEW

**Comment by kyegupov** [comment]
Kicking the Appveyor build

**Comment by kyegupov** [comment]
Kicking CI

**Comment by lili2311** [comment]
Please test locally on a example project to manually verify, how hard is it to add a test case with build flavours in it? 

It's really nice to have test cases showcase supported and unsupported features

**Comment by lili2311** [comment]
Locally tested, looks good. Just waiting to add a fixture

**Comment by kyegupov** [comment]
@lili2311 added the short error message for the top. Example analytics message now:

```snyk analytics { args:
   [ { debug: true,
       'all-sub-projects': true,
       org: undefined,
       showVulnPaths: true } ],
  command: 'bad-command',
  metadata:
   { local: true,
     error:
      'Error: \nTesting /home/kyegupov/usercode/snyk_test...\n\nGradle Error (short):\n> Could not resolve all dependencies for configuration \':app:snykMergedDepsConf\'.\n   > Could not resolve project :mylibrary.\n      > Cannot choose between the following variants of project :mylibrary:\n\n===== DEBUG INFORMATION START =====\ngradle command: \'/home/kyegupov/usercode/snyk_test/gradlew\' snykResolvedDepsJson -q --build-file build.gradle --no-daemon -Dorg.gradle.parallel= -I /tmp/tmp-22917VJwQn4QzJLRs-init.gradle\n\n------------------------------------------------------------\nGradle 5.1.1\n------------------------------------------------------------\n\nBuild time:   2019-01-10 23:05:02 UTC\nRevision:     3c9abb645fb83932c44e8610642393ad62116807\n\nKotlin DSL:   1.1.1\nKotlin:       1.3.11\nGroovy:       2.5.4\nAnt:          Apache Ant(TM) version 1.9.13 compiled on July 10 2018\nJVM:          1.8.0_212 (Oracle Corporation 25.212-b03)\nOS:           Linux 5.0.0-15-generic amd64\n\n\n\n>>> command: \'/home/kyegupov/usercode/snyk_test/gradlew\' snykResolvedDepsJson -q --build-file build.gradle --no-daemon -Dorg.gradle.parallel= -I /tmp/tmp-22917VJwQn4QzJLRs-init.gradle\n>>> exit code: 1\n>>> stdout:\nSNYKECHO snykResolvedDepsJson task is executing via doLast\nSNYKECHO processing project: snyk_test\nSNYKECHO processing project: app\nSNYKECHO processing project: mylibrary\nJSONATTRS {"org.jetbrains.kotlin.platform.type":["common","androidJvm"],"org.gradle.usage":["kotlin-api","java-runtime","android-bundle","java-api"],"com.android.build.api.attributes.BuildTypeAttr":["debug","release"],"org.jetbrains.kotlin.localToProject":["local to :app"],"com.android.build.gradle.internal.dependency.AndroidTypeAttr":["Aar","Metadata","Apk"],"com.android.build.api.attributes.VariantAttr":["debug","release"]}\nSNYKECHO constructing merged configuration from [androidApis, androidTestAnnotationProcessor, androidTestApi, androidTestApiDependenciesMetadata, androidTestApk, androidTestCompile, androidTestCompileOnly, androidTestCompileOnlyDependenciesMetadata, androidTestDebugAnnotationProcessor, androidTestDebugApi, androidTestDebugApiDependenciesMetadata, androidTestDebugApk, androidTestDebugCompile, androidTestDebugCompileOnly, androidTestDebugCompileOnlyDependenciesMetadata, androidTestDebugImplementation, androidTestDebugImplementationDependenciesMetadata, androidTestDebugProvided, androidTestDebugRuntimeOnly, androidTestDebugRuntimeOnlyDependenciesMetadata, androidTestDebugWearApp, androidTestImplementation, androidTestImplementationDependenciesMetadata, androidTestProvided, androidTestRuntimeOnly, androidTestRuntimeOnlyDependenciesMetadata, androidTestUtil, androidTestWearApp, annotationProcessor, api, apiDependenciesMetadata, apk, archives, compile, compileOnly, compileOnlyDependenciesMetadata, debugAndroidTestAnnotationProcessorClasspath, debugAndroidTestApi, debugAndroidTestApiDependenciesMetadata, debugAndroidTestCompile, debugAndroidTestCompileClasspath, debugAndroidTestCompileOnly, debugAndroidTestCompileOnlyDependenciesMetadata, debugAndroidTestImplementation, debugAndroidTestImplementationDependenciesMetadata, debugAndroidTestRuntime, debugAndroidTestRuntimeClasspath, debugAndroidTestRuntimeOnly, debugAndroidTestRuntimeOnlyDependenciesMetadata, debugAnnotationProcessor, debugAnnotationProcessorClasspath, debugApi, debugApiDependenciesMetadata, debugApiElements, debugApk, debugBundleElements, debugCompile, debugCompileClasspath, debugCompileOnly, debugCompileOnlyDependenciesMetadata, debugImplementation, debugImplementationDependenciesMetadata, debugMetadataElements, debugMetadataValues, debugProvided, debugRuntime, debugRuntimeClasspath, debugRuntimeElements, debugRuntimeOnly, debugRuntimeOnlyDependenciesMetadata, debugUnitTestAnnotationProcessorClasspath, debugUnitTestApi, debugUnitTestApiDependenciesMetadata, debugUnitTestCompile, debugUnitTestCompileClasspath, debugUnitTestCompileOnly, debugUnitTestCompileOnlyDependenciesMetadata, debugUnitTestImplementation, debugUnitTestImplementationDependenciesMetadata, debugUnitTestRuntime, debugUnitTestRuntimeClasspath, debugUnitTestRuntimeOnly, debugUnitTestRuntimeOnlyDependenciesMetadata, debugWearApp, debugWearBundling, default, implementation, implementationDependenciesMetadata, kotlinCompilerClasspath, kotlinCompilerPluginClasspath, kotlinNativeCompilerPluginClasspath, lintChecks, lintClassPath, lintPublish, provided, releaseAnnotationProcessor, releaseAnnotationProcessorClasspath, releaseApi, releaseApiDependenciesMetadata, releaseApiElements, releaseApk, releaseBundleElements, releaseCompile, releaseCompileClasspath, releaseCompileOnly, releaseCompileOnlyDependenciesMetadata, releaseImplementation, releaseImplementationDependenciesMetadata, releaseMetadataElements, releaseMetadataValues, releaseProvided, releaseRuntime, releaseRuntimeClasspath, releaseRuntimeElements, releaseRuntimeOnly, releaseRuntimeOnlyDependenciesMetadata, releaseUnitTestAnnotationProcessorClasspath, releaseUnitTestApi, releaseUnitTestApiDependenciesMetadata, releaseUnitTestCompile, releaseUnitTestCompileClasspath, releaseUnitTestCompileOnly, releaseUnitTestCompileOnlyDependenciesMetadata, releaseUnitTestImplementation, releaseUnitTestImplementationDependenciesMetadata, releaseUnitTestRuntime, releaseUnitTestRuntimeClasspath, releaseUnitTestRuntimeOnly, releaseUnitTestRuntimeOnlyDependenciesMetadata, releaseWearApp, releaseWearBundling, runtimeOnly, runtimeOnlyDependenciesMetadata, testAnnotationProcessor, testApi, testApiDependenciesMetadata, testApk, testCompile, testCompileOnly, testCompileOnlyDependenciesMetadata, testDebugAnnotationProcessor, testDebugApi, testDebugApiDependenciesMetadata, testDebugApk, testDebugCompile, testDebugCompileOnly, testDebugCompileOnlyDependenciesMetadata, testDebugImplementation, testDebugImplementationDependenciesMetadata, testDebugProvided, testDebugRuntimeOnly, testDebugRuntimeOnlyDependenciesMetadata, testDebugWearApp, testImplementation, testImplementationDependenciesMetadata, testProvided, testReleaseAnnotationProcessor, testReleaseApi, testReleaseApiDependenciesMetadata, testReleaseApk, testReleaseCompile, testReleaseCompileOnly, testReleaseCompileOnlyDependenciesMetadata, testReleaseImplementation, testReleaseImplementationDependenciesMetadata, testReleaseProvided, testReleaseRuntimeOnly, testReleaseRuntimeOnlyDependenciesMetadata, testReleaseWearApp, testRuntimeOnly, testRuntimeOnlyDependenciesMetadata, testWearApp, wearApp]\nSNYKECHO resolving configuration snykMergedDepsConf\n\n>>> stderr:\n\nFAILURE: Build failed with an exception.\n\n* Where:\nInitialization script \'/tmp/tmp-22917VJwQn4QzJLRs-init.gradle\' line: 195\n\n* What went wrong:\nExecution failed for task \':snykResolvedDepsJson\'.\n> Could not resolve all dependencies for configuration \':app:snykMergedDepsConf\'.\n   > Could not resolve project :mylibrary.\n     Required by:\n         project :app\n      > Cannot choose between the following variants of project :mylibrary:\n          - debugRuntimeElements\n          - releaseRuntimeElements\n        All of them match the consumer attributes:\n          - Variant \'debugRuntimeElements\':\n              - Found com.android.build.api.attributes.BuildTypeAttr \'debug\' but wasn\'t required.\n              - Found com.android.build.api.attributes.VariantAttr \'debug\' but wasn\'t required.\n              - Found com.android.build.gradle.internal.dependency.AndroidTypeAttr \'Aar\' but wasn\'t required.\n              - Found org.gradle.usage \'java-runtime\' but wasn\'t required.\n              - Required org.jetbrains.kotlin.localToProject \'local to :app\' but no value provided.\n          - Variant \'releaseRuntimeElements\':\n              - Found com.android.build.api.attributes.BuildTypeAttr \'release\' but wasn\'t required.\n              - Found com.android.build.api.attributes.VariantAttr \'release\' but wasn\'t required.\n              - Found com.android.build.gradle.internal.dependency.AndroidTypeAttr \'Aar\' but wasn\'t required.\n              - Found org.gradle.usage \'java-runtime\' but wasn\'t required.\n              - Required org.jetbrains.kotlin.localToProject \'local to :app\' but no value provided.\n\n* Try:\nRun with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.\n\n* Get more help at https://help.gradle.org\n\nBUILD FAILED in 7s\n\n\n===== DEBUG INFORMATION END =====\n\nError running Gradle dependency analysis.\n\nIt seems like you are scanning an Android build with ambiguous dependency variants.\nWe cannot automatically resolve dependencies for such builds.\n\nYou have several options to make dependency resolution rules more specific:\n\n1. Run Snyk CLI tool with an attribute filter, e.g.:\n    snyk test --all-sub-projects --configuration-attributes=buildtype:release,usage:java-runtime\n\nThe filter will select matching attributes from those found in your configurations, use them\nto select matching configuration(s) to be used to resolve dependencies. Any sub-string of the full\nattribute name is enough.\n\nSelect the values for the attributes that would allow to unambiguously select the correct dependency\nvariant. The Gradle error message above should contain information about attributes found in\ndifferent variants.\n\nSuggested attributes: buildtype, usage and your "flavor dimension" for Android builds.\n\nThe following attributes and their possible values were found in your configurations:\n                          org.jetbrains.kotlin.platform.type: common, androidJvm\n                                            org.gradle.usage: kotlin-api, java-runtime, android-bundle, java-api\n              com.android.build.api.attributes.BuildTypeAttr: debug, release\n                         org.jetbrains.kotlin.localToProject: local to :app\ncom.android.build.gradle.internal.dependency.AndroidTypeAttr: Aar, Metadata, Apk\n                com.android.build.api.attributes.VariantAttr: debug, release\n\n2. Run Snyk CLI tool for specific configuration(s), e.g.:\n    snyk test --gradle-sub-project=my-app --configuration-matching=\'^releaseRuntimeClasspath$\'\n\n(note that some configurations won\'t be present in every your subproject)\n\n3. Converting your subproject dependency specifications from the form of\n    implementation project(\':mymodule\')\nto\n    implementation project(path: \':mymodule\', configuration: \'default\')\n    at /home/kyegupov/code/snyk/src/cli/commands/test.ts:166:19\n    at Generator.throw (<anonymous>)\n    at rejected (/home/kyegupov/code/snyk/dist/cli/commands/test.js:5:65)\n    at process._tickCallback (internal/process/next_tick.js:68:7)',
     'error-code': undefined,
     command: 'test' },
  version:
   'master: ca26df30556aae4cf5892c97887feed0da780979 (2 dirty files)',
  os: 'Linux 5.0',
  nodeVersion: 'v10.15.2',
  id: '971d0dde8168057908c2dcf5202e686593ee9e1f',
  ci: false,
  durationMs: 8880 }```

File: lib/index.ts [hunk 1]
```diff
@@ -17,21 +17,40 @@ function debugLog(s: string) {
       debugModule.enable(process.env.DEBUG);
     }
     logger = debugModule('snyk-gradle-plugin');
-    console.log('debugger initialized ' + process.env.DEBUG);
```
Comment (lili2311) [inline lib/index.ts]: ditch please

**Pull request merged**